release-22.2: roachpb: omit some empty `RangeDescriptor` fields during marshaling

erikgrinaker opened the pull request:
**roachpb: omit some empty `RangeDescriptor` fields during marshaling**

In acdf42ad, first released in 22.2, all `RangeDescriptor` and `ReplicaDescriptor` fields were made non-nullable.  This was considered safe, because we do not rely on the range descriptor encoding being stable (the prep work for this was done back in 20.1 via ea720e3e), and we consider a `nil` value equivalent to the default zero value.

Unfortunately, `RangeDescriptor.Equal()` and `ReplicaDescriptor.Equal()` did not consider `nil` and zero values equal, unlike the rest of the code. In particular, this can prevent range config changes (e.g. moving replicas, up/downreplication, splits, merges, etc) because e.g. `execChangeReplicasTxn()` compares the caller's view of the range descriptor to its view of the canonical range descriptor stored in KV.

There are two primary scenarios where this can happen in mixed 22.1/22.2 clusters:

1. When a 22.1 leaseholder's in-memory range descriptor diverges from the canonical range descriptor in KV, because the latter was written by a 22.1 node coordinating a config change that was then committed by a past 22.2 leaseholder who propagated its view of the range descriptor to 22.1 replicas via Raft.

2. When a 22.2 node sends an `AdminChangeReplicas` request to a 22.1 leaseholder, e.g. via an `ALTER RANGE RELOCATE` statement. `AdminSplit`/`AdminMerge` are unaffected (except for 1), because they use the leaseholder's in-memory descriptor rather than the caller's.

This patch implements custom encoders for `RangeDescriptor` and `ReplicaDescriptor` which omit the `StickyBit` and `Type` fields when they have the zero value. This matches the 22.1 encoding (22.1 was careful to always normalize the zero value to `nil` due to past backwards compatibility concerns with 19.1), and preserves backwards compatibility with all 22.1 versions.

While technically a protocol change, this change is considered safe, because 22.2 nodes do not differentiate between `nil` and the zero value, and we do not rely on the wire encoding being stable. This also means that the custom encoding can be removed in 23.1 without any further backwards compatibility concerns with 22.2.

Resolves #94834.
Epic: none

Release note (bug fix): Fixed a 22.1 compatibility bug where, in a cluster with mixed 22.2/22.1 nodes, range replica changes (moving replicas, up/downreplication, splits, and merges) could sometimes fail on 22.1 leaseholders with an error of the form "change replicas of r47 failed: descriptor changed: [expected] != [actual]", without any apparent difference between the listed descriptors. This would not affect the upgrade itself, and either continuing to upgrade all nodes to 22.2 or rolling nodes back to 22.1 (possibly with an additional restart) will resolve the issue.

**roachtest: add `change-replicas/mixed-version` roachtest**

This patch adds a Roachtest which exercises replica config changes in mixed-version clusters. It does so both explicitly
with `ALTER RANGE RELOCATE` and implicitly via zone configs and the replicate queue, in several scenarios in sequence:

1. Mixed 22.1/22.2 nodes.
2. All 22.2 nodes, unfinalized.
3. All 22.1 nodes, downgraded from 22.2.
4. All 22.2 nodes, finalized.

Touches #94834.
Epic: none

Release note: None

---

Release justification: ensures backwards compatibility with 22.1 in mixed-version clusters.

/cc @cockroachdb/release

erikgrinaker committed:
roachpb: omit some empty `RangeDescriptor` fields during marshaling

In acdf42ad, first released in 22.2, all `RangeDescriptor` and
`ReplicaDescriptor` fields were made non-nullable.  This was considered
safe, because we do not rely on the range descriptor encoding being
stable (the prep work for this was done back in 20.1 via  ea720e3e), and
we consider a `nil` value equivalent to the default zero value.

Unfortunately, `RangeDescriptor.Equal()` and `ReplicaDescriptor.Equal()`
did not consider `nil` and zero values equal, unlike the rest of the
code. In particular, this can prevent range config changes (e.g. moving
replicas, up/downreplication, splits, merges, etc) because e.g.
`execChangeReplicasTxn()` compares the caller's view of the range
descriptor to its view of the canonical range descriptor stored in KV.

There are two primary scenarios where this can happen in mixed 22.1/22.2
clusters:

1. When a 22.1 leaseholder's in-memory range descriptor diverges from
the canonical range descriptor in KV, because the latter was written by
a 22.1 node coordinating a config change that was then committed by a
past 22.2 leaseholder who propagated its view of the range descriptor to
22.1 replicas via Raft.

2. When a 22.2 node sends an `AdminChangeReplicas` request to a 22.1
leaseholder, e.g. via an `ALTER RANGE RELOCATE` statement.
`AdminSplit`/`AdminMerge` are unaffected (except for 1), because they
use the leaseholder's in-memory descriptor rather than the caller's.

This patch implements custom encoders for `RangeDescriptor` and
`ReplicaDescriptor` which omit the `StickyBit` and `Type` fields when
they have the zero value. This matches the 22.1 encoding (22.1 was
careful to always normalize the zero value to `nil` due to past
backwards compatibility concerns with 19.1), and preserves backwards
compatibility with all 22.1 versions.

While technically a protocol change, this change is considered safe,
because 22.2 nodes do not differentiate between `nil` and the zero
value, and we do not rely on the wire encoding being stable. This also
means that the custom encoding can be removed in 23.1 without any
further backwards compatibility concerns with 22.2.

Epic: none

Release note (bug fix): Fixed a 22.1 compatibility bug where, in a
cluster with mixed 22.2/22.1 nodes, range replica changes (moving
replicas, up/downreplication, splits, and merges) could sometimes fail
on 22.1 leaseholders with an error of the form "change replicas of r47
failed: descriptor changed: [expected] != [actual]", without any
apparent difference between the listed descriptors. This would not
affect the upgrade itself, and either continuing to upgrade all nodes to
22.2 or rolling nodes back to 22.1 (possibly with an additional restart)
will resolve the issue.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94888)
<!-- Reviewable:end -->

erikgrinaker commented:
Marking this as ready for review. I'm working on a mixed-version roachtest covering this scenario, but will submit that separately.

erikgrinaker committed:
roachtest: add `change-replicas/mixed-version` roachtest

This patch adds a Roachtest which exercises replica config changes in
mixed-version clusters. It does so both explicitly with `ALTER RANGE
RELOCATE` and implicitly via zone configs and the replicate queue, in
several sequential scenarios:

1. Mixed 22.1/22.2 nodes.
2. All 22.2 nodes, unfinalized.
3. All 22.1 nodes, downgraded from 22.2.
4. All 22.2 nodes, finalized.

Epic: none
Release note: None

erikgrinaker commented:
I've added on a regression roachtest which exercises range config changes in mixed-version clusters, both explicitly via `ALTER RANGE RELOCATE` and implicitly via zone configs and replicate queues.

tbg commented on pkg/cmd/roachtest/tests/mixed_version_change_replicas.go:
```diff
@@ -0,0 +1,312 @@
+
+				if len(rangeErrors) > 0 {
+					for rangeID, result := range rangeErrors {
+						t.Status(fmt.Sprintf("failed to move r%d from n%d to n%d via n%d: %s",
```
Why `t.Status`? Don't you want `t.L().Printf(...)`?

erikgrinaker replied:
Just cargo culting, updated.

tbg commented on pkg/cmd/roachtest/tests/versionupgrade.go:
```diff
@@ -380,6 +380,18 @@ func uploadAndStart(nodes option.NodeListOption, v string) versionStep {
 }
 
+// uploadAndStartSequential is like uploadAndStart, but starts node sequentially
+// to bootstrap the cluster with deterministic node IDs.
```
Why don't you start from fixtures, like the other mixed-version tests? I doubt that really adds anything new for this particular test (since we're mostly operating on range descriptors split using the current version) but as the fixtures get better this test will.

erikgrinaker replied:
I though the checkpoint stuff were all about improving setup times, which isn't relevant here, but I understand now that they're used to start the clusters out with some historical state.
